08-graphs/02-graphs-algorithms.md

```diff
@@ -130,22 +234,247 @@ In the worst-case you will need to add each node to the Queue, so the space comp
+
+##### !end-question
+
+##### !placeholder
+
+```py
+class Graph:
+    
+    # The graph is stored in an adjacency dictionary where each key 
+    # represents an item in the graph and each value in the dictionary
+    # corresponds to a list of edges from the key
+    def __init__(self, adjacency_dict = {}):
+        self.adjacency_dict = adjacency_dict
+
+    def dfs(self):
+        pass
+```
+
+##### !end-placeholder
+
+##### !tests
+
+```py
+import unittest
+from main import *
+
+class TestPython1(unittest.TestCase):
+    def test_dfs(self):
+        adjacency_dict = {
+            "Seattle": ["Chicago", "Portland"],
+            "Chicago": ["Seattle"],
+            "Portland": ["Seattle", "Hawaii"],
+            "Hawaii": ["Portland", "Juneau"],
+            "Juneau": ["Hawaii"]
+        }
+
+        g = Graph(adjacency_dict)
+
+        answer = ["Seattle", "Portland", "Hawaii", "Juneau", "Chicago"]
+        self.assertEqual(answer, g.dfs())
+
+    def test_dfs_empty_graph(self):
+        g = Graph()
+        self.assertEqual([], g.dfs())
+
+    def test_dfs_one_item(self):
+        adjacency_dict = {
+            "Seattle": []
+        }
+        g = Graph(adjacency_dict)
+        self.assertEqual(["Seattle"], g.dfs())
+```
+
+##### !end-tests
+
+### !hint
 **Pseudocode**
 
-1. Starting with a graph's adjacency list and a starting node `s`
+1. Start by grabbing the first item in the adjacency dictionary `start_node`
 1. Create a Stack called `stack`
-1. Create a list of nodes called `visited` and mark each element `false`
-1. Push `s` into `stack`
-1. mark `s` as `true` in `visited`
+1. Create an empty list called `visited`
+1. Push `start_node` onto `stack`
 1. while `stack` is not empty
-    1. Remove an element from `stack` and store it in `current_node`
-    1. Loop through all the neighbors of `current_node`
-        1. If they are not marked as `true` in `visited`
-            1. Push the neighbor into `stack`
-            1. Mark the neighbor as `true` in `visited`
+    1. Pop the stack and store it in `current`
+    1. Add `current` to `visited`
+    1. Loop through all the neighbors of `current`
+        1. If they are not in `visited`
+            1. Push the neighbor onto `stack`
+1. Return `visited`
+<br>
+Still feeling stuck? Check this video walkthrough of the solution.
+
+<!-- Insert link to video explanation here -->
+
+### !end-hint 
+
+### !explanation
+An example of a working implementation:
+
+```python
+def dfs(self):
+    graph = self.adjacency_dict
+    
+    if len(graph) == 0:
+        return []
+        
+    first_item = list(graph.keys())[0]
+    stack = [first_item]
+    visited = []
+
+    while stack:
+        current = stack.pop()
+        visited.append(current)
+
+        for neighbor in graph[current]:
+            if neighbor not in visited:
+                stack.append(neighbor)
+
+    return visited
+```
+### !end-explanation
 
+### !end-challenge
 
```

Comment:
Before asking this question, consider adding in an observation that when we see a stack data structure being used (as well as we are doing the same operation over and over on different subgraphs) we may think about using recursion.